src/components/singleton-editor/reducer.ts

```diff
@@ -0,0 +1,175 @@
+import { Content } from "../../abstractions/domain/content";
+
+export type Action = Readonly<
+  | {
+      type: "content-updated";
+    }
+  | {
+      type: "content-saved";
+      savingUpdateCounter: number;
+    }
+  | {
+      type: "save-requested";
+      force: boolean;
+    }
+  | {
+      type: "content-prepared-to-save";
+      savingUpdateCounter: number;
+      content: Content;
+    }
+  | {
+      type: "when-all-saved-action-requested";
+      action: () => void;
+    }
+>;
+
+export type SavingProcessDataPreparingContent = Readonly<{
+  step: "preparing-content";
+  savingUpdateCounter: number;
+}>;
+
+export type SavingProcessDataPostingContent = Readonly<{
+  step: "posting-content";
+  savingUpdateCounter: number;
+  content: Content;
+}>;
+
+export type SavingProcessData =
+  | null
+  | SavingProcessDataPreparingContent
+  | SavingProcessDataPostingContent;
+
+export type State = Readonly<{
+  savedCounter: number;
+  updateCounter: number;
+  savingProcessData: SavingProcessData;
+  onNoPendingUpdates: null | (() => void);
+}>;
+
+export const reducerInitialState: State = {
+  savedCounter: 0,
+  updateCounter: 0,
+  savingProcessData: null,
+  onNoPendingUpdates: null,
+};
+
+export function reducer(
+  { savedCounter, updateCounter, savingProcessData, onNoPendingUpdates }: State,
+  action: Action
+): State {
+  if (
+    savingProcessData &&
+    savingProcessData.savingUpdateCounter > updateCounter
+  ) {
+    throw new Error(
+      "Unexpected scenario: savingDataCounter cannot be greater than updateCounter"
+    );
+  }
+  if (savedCounter > updateCounter) {
+    throw new Error(
+      "Unexpected scenario: savedDataCounter cannot be greater than updateCounter"
+    );
+  }
```

Comment:
No estoy muy seguro de esta parte. [Por aqui](https://stackoverflow.com/questions/62388219/is-it-safe-to-throw-an-error-in-a-react-reducer) hablan un poco al respecto indicando que debería chequearse antes. Otro dice que deberia modificar una propiedad indicando que hay un error. Nosé que piensas acerca de esto

Reply:
En realidad son escenarios inesperados, no deberían ocurrir salvo que haya un error en el código, entonces me gustaría hacerlos evidentes. 

Creo que modificar una propiedad indicando un error no sería correcto ya son escenarios teóricamente imposibles.

Igual voy a leer el artículo, y luego te digo.

Gracias!

Reply:
Leí el artículo y algunos de los links, estoy de acuerdo con ellos.

Pero entiendo que, en general, se refieren a excepciones que forman parte del flujo, por ejemplo errores de validación, que podrían ser capturadas y resueltas de alguna manera.

Las excepciones que yo puse, corresponden a escenarios imposibles que no puedo manejar con los tipos, no hay forma correcta de capturarlas, mi único interés es que si se da alguno de esos escenarios, se rompa cuanto antes de manera de identificar el problema.

Una alternativa sería usar simples `console.error`, pero entiendo que eso puede pasar desapercibido, en cambio la excepción va a generar claramente un error.